projecttacoma/fqm-execution
# Resolve missing valuesets from VSAC API

# Summary
This PR adds the ability for fqm-execution to make HTTP requests to a FHIR API to retrieve expanded valuesets that aren't included in the measure bundle. The bulk of the new resolution code is in `src/helpers/ValueSetResolver.ts`.

## New behavior
* Added `-a`/`--vs-api-key` flag to CLI, and `options.vsAPIKey` to library option object
* Added a `ValueSetResolver` to reach out to the valuesets' URLs and retrieve them (or throw an error)
* Valueset parsing now handles expanded valuesets as well as composed ones. Expansions are considered the default if both are present.

## Code changes
* Turned a bunch of functions into async functions
* Added `ValueSetResolver` class to provide valueset resolution via http, as well as checking whether all missing valuesets have been retrieved
* Moved `ValueSetResolver` and `ValueSetHelper` into a `helpers` directory
* Added `ValueSet` expansion conversion to `valueSetsForCodeService` in `ValueSetHelper`
* Moved `getMissingDependentValuesets` into `ValueSetHelper`

# Testing guidance
* run an `npm install` to get new packages
* Run the fqm-execution CLI using your preferred method, with the `test/fixtures/EXM130-7.3.000-bundle-nocodes-missingVS.json` measure bundle. Result should be an error like `Missing the following valuesets: http://cts.nlm.nih.gov/fhir/ValueSet/2.16.840.1.113883.3.464.1003.101.12.1016, and no API key was provided to resolve them`
* Get your NLM API key (instructions in the README), and provide it to the CLI using `-a <api_key>` and the above measure bundle. Result should be a successful calculation.
